### ISSUE TITLE
Support Dynamic Type/Attributes/Parameters

### ISSUE THREAD

**Issue by paddycarver** [hashicorp/terraform-plugin-framework#147]
### Module version

```
v0.3.0
```

### Use-cases

There are sometimes occasions in which provider developers want to use unstructured output or input. See, for example, the [`kubernetes_manifest`](https://github.com/hashicorp/terraform-provider-kubernetes/tree/3cf97d066d26f29744a980a8354d673d68c1a12a/manifest) implementation, or the [`tfe_outputs` data source](https://github.com/hashicorp/terraform-provider-tfe/blob/749206d33844b0c1e506e1b8c03120c31de47782/tfe/data_source_outputs.go). See also hashicorp/terraform-provider-external#76. Allowing people to do this from the framework offers a friendlier interface for this functionality than plugin-go.

### Attempted Solutions

Traditionally, this has been achieved through muxing in plugin-go.

### Proposal

If we add a `types.DynamicType` and `types.Dynamic` implementation, we can natively support this:

```go
type DynamicType struct{}

func (d DynamicType) TerraformType(_ context.Context) tftypes.Type {
  return tftypes.DynamicPseudoType
}

func (d DynamicType) ValueFromTerraform(_ context.Context, val tftypes.Value) (attr.Value, error) {
  return Dynamic{Value: val}, nil
}

// ... etc.

type Dynamic struct{
  Value tftypes.Value
}

// ... etc
```

**Comment by apparentlymart** [comment]
The design you've shown here does seem like it would neatly sidestep one of the shortcomings of the `cty` design: you've declared "dynamic" as a real type which _has a_ value of some other concrete (but dynamically-chosen) type inside of it, whereas `cty.DynamicPseudoType` is only a type _constraint_ and there are never any concrete values of that type. (type constraint vs. type confusion is a common hazard in practical use of `cty`.)

The downside of such a design is that producers and consumers of such values will presumably need to explicitly wrap and unwrap the values in the "dynamic" container, which is less convenient but more explicit than just being given a value and inspecting it directly to find out what type it has.

Overall I think your approach here is a net win in that the explicitness resolves an ambiguity in the `cty` design which commonly leads to misunderstandings and therefore bugs, and thus the extra wrapping/unwrapping pays for itself.

**Comment by m13t** [comment]
I’m interested to utilise dynamic types for an internal provider we have. If I try to create a custom type using dynamic pseudo type at the moment however, the runtime plan fails as it states types don’t match. Is there additional inner workings that need to be updated to support dynamic pseudo type constraint?

**Comment by paddycarver** [comment]
There very well may be assumptions that don't hold up in the face of DPT that we'll need to fix. We haven't exercised that code path yet. If you post your code and the error you're getting, we may be able to tell you more.

**Comment by m13t** [comment]
Sure I’ll get a concise example together but it’s mostly based on the above proposal.

**Comment by sebhoss** [comment]
I have some error & workaround to share. The error message looks like this:

```shell
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
Please report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version, the stack trace
shown below, and any additional information which may help replicate the issue.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

inconsistent list element types (cty.Object(map[string]cty.Type{"args":cty.List(cty.String), "command":cty.List(cty.String), "env":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String, "value_from":cty.Object(map[string]cty.Type{"config_map_key_ref":cty.Object(map[string]cty.Type{"key":cty.String, "name":cty.String, "optional":cty.Bool}), "field_ref":cty.Object(map[string]cty.Type{"api_version":cty.String, "field_path":cty.String}), "resource_field_ref":cty.Object(map[string]cty.Type{"container_name":cty.String, "divisor":cty.String, "resource":cty.String}), "secret_key_ref":cty.Object(map[string]cty.Type{"key":cty.String, "name":cty.String, "optional":cty.Bool})})})), "env_from":cty.List(cty.Object(map[string]cty.Type{"config_map_ref":cty.Object(map[string]cty.Type{"name":cty.String, "optional":cty.Bool}), "prefix":cty.String, "secret_ref":cty.Object(map[string]cty.Type{"name":cty.String, "optional":cty.Bool})})), "image":cty.String, "image_pull_policy":cty.String, "lifecycle":cty.Object(map[string]cty.Type{"post_start":cty.Object(map[string]cty.Type{"exec":cty.Object(map[string]cty.Type{"command":cty.List(cty.String)}), "http_get":cty.Object(map[string]cty.Type{"host":cty.String, "http_headers":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String})), "path":cty.String, "port":cty.DynamicPseudoType, "scheme":cty.String}), "tcp_socket":cty.Object(map[string]cty.Type{"host":cty.String, "port":cty.DynamicPseudoType})}), "pre_stop":cty.Object(map[string]cty.Type{"exec":cty.Object(map[string]cty.Type{"command":cty.List(cty.String)}), "http_get":cty.Object(map[string]cty.Type{"host":cty.String, "http_headers":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String})), "path":cty.String, "port":cty.DynamicPseudoType, "scheme":cty.String}), "tcp_socket":cty.Object(map[string]cty.Type{"host":cty.String, "port":cty.DynamicPseudoType})})}), "liveness_probe":cty.Object(map[string]cty.Type{"exec":cty.Object(map[string]cty.Type{"command":cty.List(cty.String)}), "failure_threshold":cty.Number, "grpc":cty.Object(map[string]cty.Type{"port":cty.Number, "service":cty.String}), "http_get":cty.Object(map[string]cty.Type{"host":cty.String, "http_headers":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String})), "path":cty.String, "port":cty.Number, "scheme":cty.String}), "initial_delay_seconds":cty.Number, "period_seconds":cty.Number, "success_threshold":cty.Number, "tcp_socket":cty.Object(map[string]cty.Type{"host":cty.String, "port":cty.DynamicPseudoType}), "termination_grace_period_seconds":cty.Number, "timeout_seconds":cty.Number}), "name":cty.String, "ports":cty.List(cty.Object(map[string]cty.Type{"container_port":cty.Number, "host_ip":cty.String, "host_port":cty.Number, "name":cty.String, "protocol":cty.String})), "readiness_probe":cty.Object(map[string]cty.Type{"exec":cty.Object(map[string]cty.Type{"command":cty.List(cty.String)}), "failure_threshold":cty.Number, "grpc":cty.Object(map[string]cty.Type{"port":cty.Number, "service":cty.String}), "http_get":cty.Object(map[string]cty.Type{"host":cty.String, "http_headers":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String})), "path":cty.String, "port":cty.Number, "scheme":cty.String}), "initial_delay_seconds":cty.Number, "period_seconds":cty.Number, "success_threshold":cty.Number, "tcp_socket":cty.Object(map[string]cty.Type{"host":cty.String, "port":cty.DynamicPseudoType}), "termination_grace_period_seconds":cty.Number, "timeout_seconds":cty.Number}), "resources":cty.Object(map[string]cty.Type{"limits":cty.Map(cty.String), "requests":cty.Map(cty.String)}), "security_context":cty.Object(map[string]cty.Type{"allow_privilege_escalation":cty.Bool, "capabilities":cty.Object(map[string]cty.Type{"add":cty.List(cty.String), "drop":cty.List(cty.String)}), "privileged":cty.Bool, "proc_mount":cty.String, "read_only_root_filesystem":cty.Bool, "run_as_group":cty.Number, "run_as_non_root":cty.Bool, "run_as_user":cty.Number, "se_linux_options":cty.Object(map[string]cty.Type{"level":cty.String, "role":cty.String, "type":cty.String, "user":cty.String}), "seccomp_profile":cty.Object(map[string]cty.Type{"localhost_profile":cty.String, "type":cty.String}), "windows_options":cty.Object(map[string]cty.Type{"gmsa_credential_spec":cty.String, "gmsa_credential_spec_name":cty.String, "host_process":cty.Bool, "run_as_user_name":cty.String})}), "startup_probe":cty.Object(map[string]cty.Type{"exec":cty.Object(map[string]cty.Type{"command":cty.List(cty.String)}), "failure_threshold":cty.Number, "grpc":cty.Object(map[string]cty.Type{"port":cty.Number, "service":cty.String}), "http_get":cty.Object(map[string]cty.Type{"host":cty.String, "http_headers":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String})), "path":cty.String, "port":cty.DynamicPseudoType, "scheme":cty.String}), "initial_delay_seconds":cty.Number, "period_seconds":cty.Number, "success_threshold":cty.Number, "tcp_socket":cty.Object(map[string]cty.Type{"host":cty.String, "port":cty.DynamicPseudoType}), "termination_grace_period_seconds":cty.Number, "timeout_seconds":cty.Number}), "stdin":cty.Bool, "stdin_once":cty.Bool, "termination_message_path":cty.String, "termination_message_policy":cty.String, "tty":cty.Bool, "volume_devices":cty.List(cty.Object(map[string]cty.Type{"device_path":cty.String, "name":cty.String})), "volume_mounts":cty.List(cty.Object(map[string]cty.Type{"mount_path":cty.String, "mount_propagation":cty.String, "name":cty.String, "read_only":cty.Bool, "sub_path":cty.String, "sub_path_expr":cty.String})), "working_dir":cty.String}) then cty.Object(map[string]cty.Type{"args":cty.List(cty.String), "command":cty.List(cty.String), "env":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String, "value_from":cty.Object(map[string]cty.Type{"config_map_key_ref":cty.Object(map[string]cty.Type{"key":cty.String, "name":cty.String, "optional":cty.Bool}), "field_ref":cty.Object(map[string]cty.Type{"api_version":cty.String, "field_path":cty.String}), "resource_field_ref":cty.Object(map[string]cty.Type{"container_name":cty.String, "divisor":cty.String, "resource":cty.String}), "secret_key_ref":cty.Object(map[string]cty.Type{"key":cty.String, "name":cty.String, "optional":cty.Bool})})})), "env_from":cty.List(cty.Object(map[string]cty.Type{"config_map_ref":cty.Object(map[string]cty.Type{"name":cty.String, "optional":cty.Bool}), "prefix":cty.String, "secret_ref":cty.Object(map[string]cty.Type{"name":cty.String, "optional":cty.Bool})})), "image":cty.String, "image_pull_policy":cty.String, "lifecycle":cty.Object(map[string]cty.Type{"post_start":cty.Object(map[string]cty.Type{"exec":cty.Object(map[string]cty.Type{"command":cty.List(cty.String)}), "http_get":cty.Object(map[string]cty.Type{"host":cty.String, "http_headers":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String})), "path":cty.String, "port":cty.DynamicPseudoType, "scheme":cty.String}), "tcp_socket":cty.Object(map[string]cty.Type{"host":cty.String, "port":cty.DynamicPseudoType})}), "pre_stop":cty.Object(map[string]cty.Type{"exec":cty.Object(map[string]cty.Type{"command":cty.List(cty.String)}), "http_get":cty.Object(map[string]cty.Type{"host":cty.String, "http_headers":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String})), "path":cty.String, "port":cty.DynamicPseudoType, "scheme":cty.String}), "tcp_socket":cty.Object(map[string]cty.Type{"host":cty.String, "port":cty.DynamicPseudoType})})}), "liveness_probe":cty.Object(map[string]cty.Type{"exec":cty.Object(map[string]cty.Type{"command":cty.List(cty.String)}), "failure_threshold":cty.Number, "grpc":cty.Object(map[string]cty.Type{"port":cty.Number, "service":cty.String}), "http_get":cty.Object(map[string]cty.Type{"host":cty.String, "http_headers":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String})), "path":cty.String, "port":cty.DynamicPseudoType, "scheme":cty.String}), "initial_delay_seconds":cty.Number, "period_seconds":cty.Number, "success_threshold":cty.Number, "tcp_socket":cty.Object(map[string]cty.Type{"host":cty.String, "port":cty.DynamicPseudoType}), "termination_grace_period_seconds":cty.Number, "timeout_seconds":cty.Number}), "name":cty.String, "ports":cty.List(cty.Object(map[string]cty.Type{"container_port":cty.Number, "host_ip":cty.String, "host_port":cty.Number, "name":cty.String, "protocol":cty.String})), "readiness_probe":cty.Object(map[string]cty.Type{"exec":cty.Object(map[string]cty.Type{"command":cty.List(cty.String)}), "failure_threshold":cty.Number, "grpc":cty.Object(map[string]cty.Type{"port":cty.Number, "service":cty.String}), "http_get":cty.Object(map[string]cty.Type{"host":cty.String, "http_headers":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String})), "path":cty.String, "port":cty.Number, "scheme":cty.String}), "initial_delay_seconds":cty.Number, "period_seconds":cty.Number, "success_threshold":cty.Number, "tcp_socket":cty.Object(map[string]cty.Type{"host":cty.String, "port":cty.DynamicPseudoType}), "termination_grace_period_seconds":cty.Number, "timeout_seconds":cty.Number}), "resources":cty.Object(map[string]cty.Type{"limits":cty.Map(cty.String), "requests":cty.Map(cty.String)}), "security_context":cty.Object(map[string]cty.Type{"allow_privilege_escalation":cty.Bool, "capabilities":cty.Object(map[string]cty.Type{"add":cty.List(cty.String), "drop":cty.List(cty.String)}), "privileged":cty.Bool, "proc_mount":cty.String, "read_only_root_filesystem":cty.Bool, "run_as_group":cty.Number, "run_as_non_root":cty.Bool, "run_as_user":cty.Number, "se_linux_options":cty.Object(map[string]cty.Type{"level":cty.String, "role":cty.String, "type":cty.String, "user":cty.String}), "seccomp_profile":cty.Object(map[string]cty.Type{"localhost_profile":cty.String, "type":cty.String}), "windows_options":cty.Object(map[string]cty.Type{"gmsa_credential_spec":cty.String, "gmsa_credential_spec_name":cty.String, "host_process":cty.Bool, "run_as_user_name":cty.String})}), "startup_probe":cty.Object(map[string]cty.Type{"exec":cty.Object(map[string]cty.Type{"command":cty.List(cty.String)}), "failure_threshold":cty.Number, "grpc":cty.Object(map[string]cty.Type{"port":cty.Number, "service":cty.String}), "http_get":cty.Object(map[string]cty.Type{"host":cty.String, "http_headers":cty.List(cty.Object(map[string]cty.Type{"name":cty.String, "value":cty.String})), "path":cty.String, "port":cty.DynamicPseudoType, "scheme":cty.String}), "initial_delay_seconds":cty.Number, "period_seconds":cty.Number, "success_threshold":cty.Number, "tcp_socket":cty.Object(map[string]cty.Type{"host":cty.String, "port":cty.DynamicPseudoType}), "termination_grace_period_seconds":cty.Number, "timeout_seconds":cty.Number}), "stdin":cty.Bool, "stdin_once":cty.Bool, "termination_message_path":cty.String, "termination_message_policy":cty.String, "tty":cty.Bool, "volume_devices":cty.List(cty.Object(map[string]cty.Type{"device_path":cty.String, "name":cty.String})), "volume_mounts":cty.List(cty.Object(map[string]cty.Type{"mount_path":cty.String, "mount_propagation":cty.String, "name":cty.String, "read_only":cty.Bool, "sub_path":cty.String, "sub_path_expr":cty.String})), "working_dir":cty.String}))
goroutine 74 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:24 +0x65
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x19
github.com/hashicorp/terraform/internal/logging.PanicHandler()
	/go/pkg/mod/github.com/hashicorp/terraform@v1.3.0-dev.0.20221020194023-730756eca23f/internal/logging/panic.go:55 +0x153
panic({0x21e7200, 0xc003e71bd0})
	/usr/local/go/src/runtime/panic.go:890 +0x262
github.com/zclconf/go-cty/cty.ListVal({0xc0086e4700, 0x2, 0xc0036d8c80?})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/value_init.go:166 +0x42e
github.com/zclconf/go-cty/cty/msgpack.unmarshalList(0x2b6f590?, {{0x2b6f600?, 0xc0036d8c80?}}, {0xc0086e46c0?, 0x4, 0x4})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:161 +0x4bc
github.com/zclconf/go-cty/cty/msgpack.unmarshal(0x221ec00?, {{0x2b6f590?, 0xc0036d8c90?}}, {0xc0086e46c0?, 0x4, 0x4})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:52 +0x41a
github.com/zclconf/go-cty/cty/msgpack.unmarshalObject(0xc0062f7f88?, 0xc0064355c0, {0xc0086e46c0?, 0x3, 0x4})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:297 +0x4b6
github.com/zclconf/go-cty/cty/msgpack.unmarshal(0x221ec00?, {{0x2b6f600?, 0xc0036d9c60?}}, {0xc0086e46c0?, 0x3, 0x4})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:60 +0x555
github.com/zclconf/go-cty/cty/msgpack.unmarshalObject(0xc0062f7f88?, 0xc0064355f0, {0xc000627e80?, 0x2, 0x2})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:297 +0x4b6
github.com/zclconf/go-cty/cty/msgpack.unmarshal(0x221ec00?, {{0x2b6f600?, 0xc0036d9c70?}}, {0xc000627e80?, 0x2, 0x2})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:60 +0x555
github.com/zclconf/go-cty/cty/msgpack.unmarshalObject(0xc0062f7f88?, 0xc006435650, {0xc0036d9ca0?, 0x1, 0x1})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:297 +0x4b6
github.com/zclconf/go-cty/cty/msgpack.unmarshal(0x221ec00?, {{0x2b6f600?, 0xc0036d9c80?}}, {0xc0036d9ca0?, 0x1, 0x1})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:60 +0x555
github.com/zclconf/go-cty/cty/msgpack.unmarshalObject(0xc0062f7f88?, 0xc0064356b0, {0x0?, 0x0, 0x0})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:297 +0x4b6
github.com/zclconf/go-cty/cty/msgpack.unmarshal(0xc0062f7f88?, {{0x2b6f600?, 0xc0036d9c90?}}, {0x0?, 0x0, 0x0})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:60 +0x555
github.com/zclconf/go-cty/cty/msgpack.Unmarshal({0xc003e33300, 0x12fa, 0x1300}, {{0x2b6f600?, 0xc0036d9c90?}})
	/go/pkg/mod/github.com/zclconf/go-cty@v1.11.1/cty/msgpack/unmarshal.go:22 +0x105
github.com/hashicorp/terraform/internal/plugin6.decodeDynamicValue(0x0?, {{0x2b6f600?, 0xc0036d9c90?}})
	/go/pkg/mod/github.com/hashicorp/terraform@v1.3.0-dev.0.20221020194023-730756eca23f/internal/plugin6/grpc_provider.go:688 +0x97
github.com/hashicorp/terraform/internal/plugin6.(*GRPCProvider).PlanResourceChange(_, {{0xc000056db0, 0x18}, {{{0x2b6f600, 0xc007dd2380}}, {0x0, 0x0}}, {{{0x2b6f600, 0xc00893fa10}}, {0x221f740, ...}}, ...})
	/go/pkg/mod/github.com/hashicorp/terraform@v1.3.0-dev.0.20221020194023-730756eca23f/internal/plugin6/grpc_provider.go:466 +0x859
github.com/hashicorp/terraform/internal/terraform.(*NodeAbstractResourceInstance).plan(0xc004c683c0, {0x2b85ad8, 0xc004754380}, 0x0, 0x0, 0x0, {0x0, 0x0, 0x0?})
	/go/pkg/mod/github.com/hashicorp/terraform@v1.3.0-dev.0.20221020194023-730756eca23f/internal/terraform/node_resource_abstract_instance.go:802 +0x12a3
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstance).managedResourceExecute(0xc0089de080, {0x2b85ad8, 0xc004754380})
	/go/pkg/mod/github.com/hashicorp/terraform@v1.3.0-dev.0.20221020194023-730756eca23f/internal/terraform/node_resource_plan_instance.go:218 +0xd6e
github.com/hashicorp/terraform/internal/terraform.(*NodePlannableResourceInstance).Execute(0x0?, {0x2b85ad8?, 0xc004754380?}, 0x0?)
	/go/pkg/mod/github.com/hashicorp/terraform@v1.3.0-dev.0.20221020194023-730756eca23f/internal/terraform/node_resource_plan_instance.go:60 +0x90
github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc004c660f0, {0x2b85ad8, 0xc004754380}, {0x7f08101c3640, 0xc0089de080})
	/go/pkg/mod/github.com/hashicorp/terraform@v1.3.0-dev.0.20221020194023-730756eca23f/internal/terraform/graph_walk_context.go:136 +0xc2
github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x25706a0, 0xc0089de080})
	/go/pkg/mod/github.com/hashicorp/terraform@v1.3.0-dev.0.20221020194023-730756eca23f/internal/terraform/graph.go:75 +0x315
github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc0089e0180, {0x25706a0, 0xc0089de080}, 0xc0089de1c0)
	/go/pkg/mod/github.com/hashicorp/terraform@v1.3.0-dev.0.20221020194023-730756eca23f/internal/dag/walk.go:381 +0x2f6
created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update
	/go/pkg/mod/github.com/hashicorp/terraform@v1.3.0-dev.0.20221020194023-730756eca23f/internal/dag/walk.go:304 +0xf65}
```

This is happening because of the following example terraform config (simplified):

```hcl
resource "example_resource" "testcase" {
  list_of_objects = [
    {
      name = "first"
      dynamic_attribute = 123
    },
    {
      name = "second"
    },
  ]
}
```

`dynamic_attribute` in this example is using `tftypes.DynamicPseudoType` and the long-ish error message basically says that there is a type inconsistency between the first object in the list (`cty.Number`) and the second one (`cty.DynamicPseudoType`). The problem disapears once I specify `dynamic_attribute` for the second object as well.

I think the fundamental problem here is that `TerraformType(_ context.Context)` always returns `tftypes.DynamicPseudoType` whereas `ToTerraformValue(_ context.Context)` returns the actual value & type that is used in the configuration.

**Comment by apparentlymart** [comment]
Thanks for sharing that, @sebhoss!

I expect there is an improvement to make to the framework to avoid a value like this getting so far through before it is detected, but what you showed here also indicates a bug in Terraform Core: it seems to be possible for a provider to crash Terraform Core by returning a value that doesn't conform to the schema, whereas that's supposed to be handled as a proper error message so that Terraform Core will at least get a chance to exit cleanly and tidy up any other concurrent operations in progress.

Would you mind reporting this in the `hashicorp/terraform` repository too, so that the Terraform Core team can look into that part of the problem more closely? If you do, it'd be helpful to share the relevant part of the schema of the provider (as produced by `terraform providers schema -json`) so we can see exactly what type constraint the provider is effectively declaring (after stripping away the framework abstractions). Thanks!

**Comment by bflad** [comment]
Please note that the framework does not officially support `tftypes.DynamicPseudoType` currently. It may be possible to create custom types with `tftypes.DynamicPseudoType` via the `attr.Type` and `attr.Value` interfaces, but there has not been any effort in testing it, nor building any validation or other potential features in the framework to natively support that special type. The framework type system is currently tailored for 1:1 mapping of "schema" type to value type, so there would need to be some investigation and design effort to handle how dynamic typing can interact in that model.

**Comment by sebhoss** [comment]
@apparentlymart ok created https://github.com/hashicorp/terraform/issues/32092 for that

@bflad yeah I understand. I was aware of this ticket here before deciding to give it a try and was actually surprised how well it all went. In most cases it just works(tm) and I only shared the log output as input for future design work in case you guys ever decide to implement a dynamic type. If there is anything I can do to help or try out, please let me know!

**Comment by tyler-dunkel** [comment]
@bflad thank you for pointing me to this issue. Just trying to catch up here, it looks like using `DPT` might work (although hasnt been verified from what I can see) for now but is there any anticipation on when this might be better supported natively in the provider framework?

**Comment by bflad** [comment]
Please note that we will also use this issue as a catching point for dynamic function parameter support. This functionality is being actively considered for implementation in the coming weeks (but not fully committed yet).

**Comment by austinvalle** [comment]
If you've been following along in this issue, Dynamic type support has just landed in [`terraform-plugin-framework v1.7.0`](https://github.com/hashicorp/terraform-plugin-framework/releases/tag/v1.7.0) 🥳. This includes dynamic attributes (resource, data source, and provider schemas) and dynamic function parameters/returns.

We have written some initial documentation to capture some of the scenarios you'll need to account for when using a dynamic type: https://developer.hashicorp.com/terraform/plugin/framework/handling-data/dynamic-data _(may need to let our CDN catch up, since it's just been released)_.

As always, if you run into any problems, please submit a new issue! 
https://github.com/hashicorp/terraform-plugin-framework/issues
